### ISSUE TITLE
Dealing with DateTime and DateTimeOffset

### ISSUE THREAD

**Issue by ili** [RSuter/NSwag#1949]
In our case we do use two Date Time types: DateTime (in the case when we do not care on timezone) & DateTimeOffset (when it matters).

Now we are using TS Angular template with Date as Date Time Type setting.

The problem is: both this types are serialized in the same manner - they are send as UTC time. 

For example, i'n generating report from `2019-01-01 12:00` to `2019-01-01 13:00` and i'm in +5 timezone, so server receives `2019-01-01T07:00:00.000Z` to `2019-01-01T08:00:00.000Z`

When deserialization is done to DateTimeOffset (or DateTime) it would be presented in **server's** timezone, for ex, if my server is in +3, i'll receive in controller from `2019-01-01 10:00` to `2019-01-01 11:00`. And it is critical either for DateTime, or for DateTimeOffset - because we do have wrong time.

From server side i can customize serialization to avoid this behaviour, but it should be proper sent from browser:
 * `YYYY-MM-DDTHH:MM:SS` - for DateTime types
 * `YYYY-MM-DDTHH:MM:SS+TZ:TZ` for DateTimeOffset types

In other words i do need to owerride `toString` && `toJSON` methods. 

Are there any possibility to do it now, if no, please give me advice how to implement it, and i'll look forward to make PR

**Comment by ili** [comment]
This is aldo related with #1007 (the same for C#)

**Comment by RSuter** [comment]
Have you looked into using MomentJS?

**Comment by ili** [comment]
Truly say I'm not falmiliar with MomentJS, but looking documentation I can't see the way to control serialization.

**Comment by Melfi11** [comment]
We´ve got the same situation like described above.
So we tried to change to MomentJS without success.

The solution is to use "OffsetMomentJS" in nswag configuration. Then the generated initialize method uses the parseZone functionality of MomentJS to get the corret locale date.

![image](https://user-images.githubusercontent.com/38665702/53005392-76a4d080-3433-11e9-9da9-f0abcec24c10.png)

In case of using Date or MomentJS the client will take the date given from our API endpoint and creates an date object with locale timezone but without translating the time to this timezone.

**Comment by ranouf** [comment]
@Melfi11 Hi, thanks for your help, where can I set 

> "keepOffset" = true

I would like to avoid to update the file generated by NswagStudio, I could forget to update it after each update.

**Comment by RSuter** [comment]
So how should we proceed? Can you show the current output and the expected output?

**Comment by Melfi11** [comment]
@ranouf - I`m not changing the code generated by nswag studio. In all projects we have wrapped functionalities calling the generated client from nswag studio (this prevents high efforts if there are changes in API routes or something else). In this wrapped functions I make sure that all datetime objects are converted to UTC. So I can handle the current client with setting "OffsetMomentJS".

@RSuter I know that dealing datetime objects with different timezones are a mess in each project. So I am not sure if we should declare this as bug. I think it should be fine if you are documenting your time information input and output at the client based on time zones and standard behavior of MomentJS and the JS standard Date format.

**Comment by ranouf** [comment]
@Melfi11 
Could you share a short sample of how you do that? Not sure to see how I can do that.

**Comment by Melfi11** [comment]
```ts
async get() {
    //calling get method generated by NSWAG client
    let resp = await this.NSWAGClient.getMethod();
    //here we have nothing to do
    //resp.date will be converted to time of locale timezone if we are using "OffsetMomentJS" as date default handling
    return resp;
}

async post(request) {
    //posting a request object including date object(s)
    //we will send datetime objects as utc to Web API
    //we have to convert time of object from timezone to utc
    request.date = request.date.add(-request.date.utcOffset(), 'm').utc();
    //calling post method generated by NSWAG client
    let resp = await this.NSWAGClient.post(request);
    return resp;
}
```
@ranouf - I hope I explained it clear enough?

**Comment by ili** [comment]
Coming back to this topick, just to share my dirty hack solution (easy to apply because we are using Angular SPA):

```js
export function createTimeStringFromDate(date: Date): string {
  const h = (date.getHours().toString().length === 1) ? '0' + date.getHours().toString() : date.getHours().toString();
  const m = (date.getMinutes().toString().length === 1) ? '0' + date.getMinutes().toString() : date.getMinutes().toString();
  const s = (date.getSeconds().toString().length === 1) ? '0' + date.getSeconds().toString() : date.getSeconds().toString();

  return `${h}:${m}:${s}`;
}

export function createDateStringFromDate(date: Date): string {
    return date.getFullYear() + '-' +
      (date.getMonth() < 9 ? ('0' + (date.getMonth() + 1)) : (date.getMonth() + 1)) + '-' +
      (date.getDate() < 10 ? ('0' + date.getDate()) : date.getDate());
}

export function toDateTimeString(date: Date) {
    return `${createDateStringFromDate(date)}T${createTimeStringFromDate(date)}`;
}

export function createOffsetString(date: Date): string {
    const offset = date.getTimezoneOffset();
    const offsetHours = Math.abs(offset) / 60;
    const offsetMinutes = Math.abs(offset) % 60;
    const offsetSign = offset < 0 ? '+' : '-';

    return `${offsetSign}${offsetHours >= 10 ? '' : '0'}${offsetHours}:${offsetMinutes >= 10 ? '' : '0'}${offsetMinutes}`;
}
```

And in `main.ts`:

```js
Date.prototype['toISOString'] = function () { return toISODateTimeOffsetString(this); };
```

From the server side JsonConverter applied:
```cs
	public class DateTimeJsonConverter : JsonConverter
	{
		public override bool CanConvert(Type objectType)
		{
			return objectType == typeof(DateTime)
				|| objectType == typeof(DateTime?);
		}

		public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
		{
			if (reader.TokenType == JsonToken.Null)
			{
				if (objectType != typeof(DateTime?))
					throw new JsonSerializationException(string.Format("Cannot convert null value to {0}.", objectType));

				return null;
			}

			if (reader.TokenType == JsonToken.String)
			{
				var str = reader.Value.ToString();
				if (string.IsNullOrWhiteSpace(str))
				{
					if (objectType != typeof(DateTime?))
						throw new JsonSerializationException(string.Format("Cannot convert empty value to {0}.", objectType));

					return null;
				}

				if (!DateTimeOffset.TryParse(str, CultureInfo.InvariantCulture, DateTimeStyles.None, out var time)
					&& !DateTimeOffset.TryParse(str, out time))
				{
					throw new JsonSerializationException(
					string.Format("{0}: {1} is not a valid date", reader.Path, str));
				}

				
				return time.DateTime;
			}

			if (reader.TokenType == JsonToken.Date)
			{
				if (reader.DateParseHandling == DateParseHandling.DateTimeOffset)
					return ((DateTimeOffset)reader.Value).DateTime;

				return reader.Value;
			}

			throw new JsonSerializationException($"Unexpected token type: {reader.TokenType}");
		}

		public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
		{
			if (value == null)
			{
				writer.WriteNull();
			}
			else
			{
				var ts = (DateTime)value;
				//writer.WriteValue(ts.ToUniversalTime().ToString(@"yyyy-MM-ddTHH:mm:ss.fffZ"));
				writer.WriteValue(ts.ToString(@"yyyy-MM-ddTHH:mm:ss.fff"));
			}
		}
	}
```

And model binder:
```cs
	internal class DateTimeModelBinderProvider : IModelBinderProvider
	{
		public IModelBinder GetBinder(ModelBinderProviderContext context)
		{
			if (context.Metadata.ModelType == typeof(DateTime) || context.Metadata.ModelType == typeof(DateTime?))
			{
				return new DateTimeModelBinder();
			}

			return null;
		}

		private class DateTimeModelBinder : IModelBinder
		{
			public Task BindModelAsync(ModelBindingContext bindingContext)
			{
				if (bindingContext == null)
				{
					throw new ArgumentNullException(nameof(bindingContext));
				}

				var modelName = bindingContext.ModelName;
				var targetType = bindingContext.ModelType;

				// Try to fetch the value of the argument by name
				var valueProviderResult =
					bindingContext.ValueProvider.GetValue(modelName);

				if (valueProviderResult == ValueProviderResult.None)
				{
					bindingContext.Result = targetType == typeof(DateTime?)
						? ModelBindingResult.Success(null)
						: ModelBindingResult.Failed();

					return Task.CompletedTask;
				}

				bindingContext.ModelState.SetModelValue(modelName,
					valueProviderResult);

				var value = valueProviderResult.FirstValue;

				// Check if the argument value is null or empty
				if (string.IsNullOrEmpty(value))
				{
					bindingContext.Result = targetType == typeof(DateTime?)
						? ModelBindingResult.Success(null)
						: ModelBindingResult.Failed();
					return Task.CompletedTask;
				}

				if (!DateTimeOffset.TryParse(value, CultureInfo.InvariantCulture, DateTimeStyles.None, out var time)
					&& !DateTimeOffset.TryParse(value, out time))
				{
					// Non-integer arguments result in model state errors
					bindingContext.ModelState.TryAddModelError(
											modelName,
											string.Format("{0} is not a valid date", value));
					return Task.CompletedTask;
				}

				// Model will be null if not found, including for 
				// out of range id values (0, -3, etc.)
				bindingContext.Result = ModelBindingResult.Success(time.DateTime);
				return Task.CompletedTask;
			}
		}
	}
```

Startup.cs:

```
			services.AddMvc(options =>
				{
					options.ModelBinderProviders.Insert(0, new DateTimeModelBinderProvider());
				})
				.AddJsonOptions(opt =>
				{
					opt.SerializerSettings.DateParseHandling = DateParseHandling.DateTimeOffset;
					opt.SerializerSettings.Converters.Insert(0, new DateTimeJsonConverter());
				})
```

**Comment by RicoSuter** [comment]
What options do you propose to add then?

**Comment by ili** [comment]
> What options do you propose to add then?

Sorry, your question is little bit unclear :-(

**Comment by dsheng3** [comment]
Is this already possible or any updates on defining different date time formats for the TypeScript client?

I'm looking to get different Date serialization e.g. `YYYY-MM-DDTHH:MM:SS`, `YYYY-MM-DDTHH:MM:SS+TZ:TZ` or `YYYY-MM-DD`, depending on the Type in my C# backend  where I use NodaTime `LocalDateTime`, `ZonedDateTime`, `LocalDate`.

When I look in the generated Typescript code, I find something like 

`data["mydate"] = this.mydate? this.mydate.toISOString() : <any>undefined;`

Is it possible to customize this (and maybe also for the query/path parameters which seems to use the `toJSON()` method (see #2479)?

**Comment by zipacna** [comment]
thisIssue.ressurect();
After a fair bit of debugging (within my application) and research on the Issue (from the perspective of using NSwag), I would like a Configuration Option (Date Time Types `Date (implicit UTC)`), which treats incoming DateTime Fields of a C# DTO (implicit UTC) as plain strings, appends the 'Z' and runs it through Date.

Example:
1. DB (C# DateTime.UtcNow, implizit UTC): 2023-06-07 07:31:13.764290
2. Current Result (via NSwag, JSON between C# and TypeScript): "2023-06-07T07:31:13.76429"
3. NSwag Config - TypeScript Client - DTO Types - Date Time Type 'Date': 
    - Problem: new Date('2023-06-07 07:31:13.764290') // Wed Jun 07 2023 07:31:13 GMT+0200 (Central European Summer Time)
    - Fix: new Date('2023-06-07 07:31:13.764290Z') // Wed Jun 07 2023 09:31:13 GMT+0200 (Central European Summer Time)
4. Goal, from JSON to TypeScript `new Date(stringImplicitUTC + 'Z')`: 
    - Before: '2023-06-07T07:31:13.76429'
    - After: '2023-06-07T07:31:13.76429Z' // New Date of that results in 09:31 (correct local time)

Current Workaround: Backend (C#) saves Timestamp as DateTime.UtcNow to the Database (ex. 2023-06-07 09:05:20.785255) and NSwagStudio has the DTO Configuration Option 'String' selected for Date Time Types (resulting in String Fields for the Timestamp on the TypeScript Types). Only when the End User is supposed to have the DateTime displayed/calculated
- in their local time (ex. UTC+02:00) 
- is the timestamp string (implicit UTC, "2023-06-07T09:05:20.785255") 
- parsed as a Date (js/ts '2023-06-07T09:05:20.785255Z'), 
- which would be 11:05+02:00 (correct local time). For simplicity as Angular Pipe:
`transform(valueImplicitUTC: string | undefined): Date { return new Date(valueImplicitUTC + 'Z') }`

References
- Configuration Option (mix of Date and String): maybe 'Date (implicit UTC)', @Melfi11 mentioned a similar logic for MomentJS [above](https://github.com/RicoSuter/NSwag/issues/1949#issuecomment-465061207)
- Appending 'Z': as mentioned [on StackOverflow](https://stackoverflow.com/a/40768745)
- "TimeZone handling in .NET is not optimal": as mentioned on this [other SO Answer](https://stackoverflow.com/a/12483096) and detailed on yet [another SO Answer](https://stackoverflow.com/a/73807751)
